File: mvp_app/MyProject/libtsparser/src/main/java/com/tosmart/xiexuming/libtsparser/tspacket/TsPacket.java
package com.tosmart.xiexuming.libtsparser.tspacket;

import androidx.annotation.NonNull;

import java.util.Arrays;

/**
 * @author xiexuming
 */
public class TsPacket {
    public static final byte CHECK_RIGHT_BIT = 0x01;
    public static final byte CHECK_RIGHT_TWO_BITS = 0x03;
    public static final byte CHECK_RIGHT_EIGHT = 0x0F;
    public static final byte REMOVE_LEFT_THREE_BIT = 0x1F;

    private int pid;
    private int transportScramblingControl;
    private int transportErrorIndicator;
    private int payloadUnitStartIndicator;
    private int priority;
    private int adaptationFieldControl;
    private int continuityCounter;
    private byte[] payload;


    public byte[] getPayload() {
        return payload;
    }

    @NonNull
    @Override
    public String toString() {
        return "TsPacket{" +
                "pid=" + pid +
                ", transportScramblingControl=" + transportScramblingControl +
                ", transportErrorIndicator=" + transportErrorIndicator +
                ", payloadUnitStartIndicator=" + payloadUnitStartIndicator +
                ", priority=" + priority +
                ", adaptationFieldControl=" + adaptationFieldControl +
                ", continuityCounter=" + continuityCounter +
                ", payload=" + Arrays.toString(payload) +
                '}';
    }

    public void setPid(int pid) {
        this.pid = pid;
    }

    public void setObjectByByte(byte[] tsPacket) {
        int pid = (tsPacket[1] & REMOVE_LEFT_THREE_BIT) << 8 | (tsPacket[2] & 0XFF);
        this.setTransportErrorIndicator(tsPacket[1] >> 7);
        this.setPayloadUnitStartIndicator((tsPacket[1] >> 6 & CHECK_RIGHT_BIT));
        this.setPriority(tsPacket[1] >> 5 & CHECK_RIGHT_BIT);
        this.setPid(pid);
        this.setTransportScramblingControl(((tsPacket[3] >> 6) & CHECK_RIGHT_TWO_BITS));
        this.setAdaptationFieldControl(tsPacket[3] >> 4 & CHECK_RIGHT_TWO_BITS);
        this.setContinuityCounter(tsPacket[3] & CHECK_RIGHT_EIGHT);
        this.setPayload(tsPacket);
    }


    public int getPid() {
        return pid;
    }

    public int getTransportScramblingControl() {
        return transportScramblingControl;
    }

    public int getTransportErrorIndicator() {
        return transportErrorIndicator;
    }

    public void setTransportErrorIndicator(int transportErrorIndicator) {
        this.transportErrorIndicator = transportErrorIndicator;
    }

    public int getPayloadUnitStartIndicator() {
        return payloadUnitStartIndicator;
    }

    public void setPayloadUnitStartIndicator(int payloadUnitStartIndicator) {
        this.payloadUnitStartIndicator = payloadUnitStartIndicator;
    }

    public int getPriority() {
        return priority;
    }

    public int getAdaptationFieldControl() {
        return adaptationFieldControl;
    }

    public void setAdaptationFieldControl(int adaptationFieldControl) {
        this.adaptationFieldControl = adaptationFieldControl;
    }

    public int getContinuityCounter() {
        return continuityCounter;
    }

    public void setTransportScramblingControl(int transportScramblingControl) {
        this.transportScramblingControl = transportScramblingControl;
    }

    public void setPriority(int priority) {
        this.priority = priority;
    }

    public void setContinuityCounter(int continuityCounter) {
        this.continuityCounter = continuityCounter;
    }

    public void setPayload(byte[] data) {
        byte[] payload = new byte[data.length - 4];
        System.arraycopy(data, 4, payload, 0, payload.length);
        this.payload = payload;
    }
}

File: newpaserapp/libtosmartparse/src/main/java/com/example/libtosmartparse/packet/TsPacket.kt
package com.example.libtosmartparse.packet

import java.util.*
import kotlin.experimental.and

/**
 * @author xxm
 * 2022/7/17
 **/
class TsPacket {
    private var pid = 0
    private var transportScramblingControl = 0
    private var transportErrorIndicator = 0
    private var payloadUnitStartIndicator = 0
    private var priority = 0
    private var adaptationFieldControl = 0
    private var continuityCounter = 0
    private lateinit var payload: ByteArray

    fun getPid(): Int {
        return pid
    }

    fun getTransportScramblingControl(): Int {
        return transportScramblingControl
    }

    fun getTransportErrorIndicator(): Int {
        return transportErrorIndicator
    }

    fun getPayloadUnitStartIndicator(): Int {
        return payloadUnitStartIndicator
    }

    fun getPriority(): Int {
        return priority
    }

    fun getAdaptationFieldControl(): Int {
        return adaptationFieldControl
    }

    fun getContinuityCounter(): Int {
        return continuityCounter
    }

    fun getPayload(): ByteArray {
        return payload
    }

    override fun toString(): String {
        return "TsPacket{" +
                "pid=" + pid +
                ", transportScramblingControl=" + transportScramblingControl +
                ", transportErrorIndicator=" + transportErrorIndicator +
                ", payloadUnitStartIndicator=" + payloadUnitStartIndicator +
                ", priority=" + priority +
                ", adaptationFieldControl=" + adaptationFieldControl +
                ", continuityCounter=" + continuityCounter +
                ", payload=" + Arrays.toString(payload) +
                '}'
    }

    fun setObjectByByte(tsPacket: ByteArray) {
        pid =
            ((tsPacket[1] and REMOVE_LEFT_THREE_BIT) shl 8) or (tsPacket[2] and 0xFF)
        transportErrorIndicator = tsPacket[1] shr 7
        payloadUnitStartIndicator = tsPacket[1] shr 6 and CHECK_RIGHT_BIT
        priority = tsPacket[1] shr 5 and CHECK_RIGHT_BIT
        transportScramblingControl = (tsPacket[3] shr 6) and CHECK_RIGHT_TWO_BITS
        adaptationFieldControl = tsPacket[3] shr 4 and CHECK_RIGHT_TWO_BITS
        continuityCounter = (tsPacket[3] and CHECK_RIGHT_EIGHT).toInt()

        setPayload(tsPacket)
    }

    fun setPayload(data: ByteArray) {
        this.payload = ByteArray(data.size - 4)
        System.arraycopy(data, 4, this.payload, 0, data.size - 4)
    }

    companion object {
        const val CHECK_RIGHT_BIT: Byte = 0x01
        const val CHECK_RIGHT_TWO_BITS: Byte = 0x03
        const val CHECK_RIGHT_EIGHT: Byte = 0x0F
        const val REMOVE_LEFT_THREE_BIT: Byte = 0x1F
    }
}



